crates/plex-api/src/server/library.rs

```diff
@@ -125,6 +127,113 @@ where
     }
 }
 
+/// A single media format for a `MediaItem`.
+pub struct Media<'a> {
+    client: &'a HttpClient,
+    media: &'a MediaMetadata,
+}
+
+impl<'a> Media<'a> {
+    /// The different parts that make up this media. They should be played in
+    /// order.
+    pub fn parts(&self) -> Vec<Part> {
+        self.media
+            .parts
+            .iter()
+            .map(|part| Part {
+                client: self.client,
+                part,
+            })
+            .collect()
+    }
+
+    /// The internal metadata for the media.
+    pub fn metadata(&self) -> &MediaMetadata {
+        self.media
+    }
+}
+
+/// One part of a `Media`.
+pub struct Part<'a> {
+    client: &'a HttpClient,
+    part: &'a PartMetadata,
+}
+
+impl<'a> Part<'a> {
+    /// The length of this file on disk in bytes.
+    pub fn len(&self) -> Option<u64> {
+        self.part.size
+    }
+
+    /// Downloads the original media file for this part. Optionally a range of bytes within the
+    /// file can be requested allowing for resumable transfers.
+    pub async fn download<R: RangeBounds<u64>>(&self, range: R) -> Result<AsyncBody> {
+        let path = format!("{}?download=1", self.part.key);
+
+        let start = match range.start_bound() {
+            std::ops::Bound::Included(v) => *v,
+            std::ops::Bound::Excluded(v) => v + 1,
+            std::ops::Bound::Unbounded => 0,
+        };
+
+        let end = match range.end_bound() {
+            std::ops::Bound::Included(v) => Some(*v),
+            std::ops::Bound::Excluded(v) => Some(v - 1),
+            std::ops::Bound::Unbounded => None,
+        };
+
+        let builder = self.client.get(path);
+
+        if start == 0 && (end.is_none() || end == self.part.size) {
+            // We're requesting the entire file.
+            let response = builder.send().await?;
+
+            match response.status() {
+                StatusCode::OK => Ok(response.into_body()),
+                _ => Err(crate::Error::from_response(response).await),
+            }
+        } else {
+            // We're requesting part of the file.
+            let end = end.map(|v| v.to_string()).unwrap_or_default();
+            let response = builder
+                .header("Range", format!("bytes={start}-{end}"))
+                .send()
+                .await?;
+
+            match response.status() {
+                StatusCode::PARTIAL_CONTENT => Ok(response.into_body()),
+                _ => Err(crate::Error::from_response(response).await),
+            }
+        }
```

Comment:
How do you feel about shrinking the code a tiny bit?

```suggestion
        let mut builder = self.client.get(path);

        if start != 0 || (!end.is_none() && end != self.part.size) {
            // We're requesting part of the file.
            let end = end.map(|v| v.to_string()).unwrap_or_default();
            builder = builder.header("Range", format!("bytes={start}-{end}"))
        }

        let response = builder.send().await?;

        match response.status() {
            StatusCode::OK | StatusCode::PARTIAL_CONTENT => Ok(response.into_body()),
            _ => Err(crate::Error::from_response(response).await),
        }
```